Consul template structure and first content (#13)
boruszak/consul-ia-experiments: website/content/docs/automate/consul-template/mode.mdx

---
layout: docs
page_title: Consul Template modes
description: >-
  Consul Template is a tool available as a distinct binary that enables dynamic application configuration and secrets rotation for Consul deployments based on Go templates.
---

# Consul Template modes

Consul Template can run in different modes that changes its runtime behavior and process lifecycle.

- [Once Mode](#once-mode)
- [De-Duplication Mode](#de-duplication-mode)
- [Exec Mode](#exec-mode)

## Once Mode

When running in Once Mode, Consul template will execute each template exactly once and exit.

In Once mode, Consul Template will wait for all dependencies to be rendered. If
a template specifies a dependency (a request) that does not exist in Consul,
once mode will wait until Consul returns data for that dependency. Please note
that "returned data" and "empty data" are not mutually exclusive.

To run in Once mode, include the `-once` flag or enable in the [configuration file](/consul/docs/reference/consul-template/configuration).

> When you query for all healthy services named "foo" (`{{ service "foo" }}`), you
> are asking Consul - "give me all the healthy services named foo". If there are
> no services named foo, the response is the empty array. This is also the same
> response if there are no _healthy_ services named foo.

Author: ```suggestion
When you run the query `{{ service "foo" }}` to return all healthy services named "foo", you
are asking Consul to return all the healthy services named "foo." If there are
no services with that name, the response is the empty array. This response is identical to the
response if there are no _healthy_ services named "foo."
```

Consul template processes input templates multiple times, since the first result
could impact later dependencies:

```go
{{ range services }}
{{ range service .Name }}
{{ end }}
{{ end }}
```

In this example, we have to process the output of `services` before we can
lookup each `service`, since the inner loops cannot be evaluated until the outer
loop returns a response. Consul Template waits until it gets a response from
Consul for all dependencies before rendering a template. It does not wait until
that response is non-empty though.

<Note>

Once mode implicitly disables any wait/quiescence timers specified in configuration files or passed on the command line.

</Note>

## De-Duplication Mode

Consul Template works by parsing templates to determine what data is needed and then watching Consul for any changes to that data. This allows Consul Template to efficiently re-render templates when a change occurs. However, if there are many instances of Consul Template rendering a common template there is a linear duplication of work as each instance is querying the same data.

> To make this pattern more efficient Consul Template supports work de-duplication across instances. This can be enabled with the `-dedup` flag or via the top level [`deduplicate` configuration block](/consul/docs/reference/consul-template/configuration#de-duplication-mode). Once enabled, Consul Template uses leader election on a per-template basis to have only a single node perform the queries. Results are shared among other instances rendering the same template by passing compressed data through the Consul K/V store.

Author: ```suggestion
To make this pattern more efficient, Consul Template supports work de-duplication across instances. You can enable this feature with the `-dedup` flag or in the top-level [`deduplicate` configuration block](/consul/docs/reference/consul-template/configuration#de-duplication-mode). Once enabled, Consul Template uses leader election on a per-template basis so that only a single node performs the queries. Results are shared among other instances rendering the same template by passing compressed data through the Consul K/V store.
```
